### PR TITLE
[TF][Relay] TensorFlow Frontend support with shared params

### PR DESCRIPTION
**Pull request by lfengad**
We observe that in lots of tensorflow models used in our production environment, some operators share the same node as their attributes. For example, there might be two "tf.reduce_mean" operators each with its own "axis" attribute. Their "axis" attributes can take the same "tf.constant" node. In this case, different operators share the same node as attributes. 

However, in the current tensorflow frontend of relay, such sharing of the same param among different operators is not supported. For example, in the current implementation of "_get_param" function in "python/tvm/relay/frontend/tensorflow.py", the corresponding param will be deleted permanently using "pop" from the "_params" dictionary after being checked in "_get_param" function. In this case, if another operator needs to check the same param as its attribute, that param will not be found. Hence, the current implementation of the tensorflow frontend cannot support param sharing among different operators.   

In order to support the case of param sharing among operators, we modify the tensorflow frontend. We leverage on the python decorator method. Everytime "_get_param" is called, instead of deleting the corresponding visited param as in the current implementation, we record it into a "used_params" dictionary and keep it in the original "_params" dictionary, so that another operator can still visit it in the following process. Then after the whole relay graph completes building, we delete the params have been visited in "used_params" from "_params" to generate the final "_params" output.

Compared to the old implementation ,our new implementation doesn't delete the visited params immediately in "_get_param", but records them and deletes them at the final step. In this way, the case where different operators share the same param can be supported.  
 
Any suggestions are welcome! @tqchen @FrozenGene

### CODE REVIEW

**Comment by srkreddy1238** [comment]
yes, this is an issue I too see recently while working on TF 2.0 upgrade. Few suggestions on this.

- I think we can just remove the pop of used param and leave it assuming relay build process removes unused params automatically. cc @tqchen or @jroesch to confirm this behaviour.
- You may move the test case to default test_forward.py instead of a new file for it.

File: python/tvm/relay/frontend/tensorflow.py [hunk 1]
```diff
@@ -77,10 +77,21 @@ def _dim_check(attrs):
         return False
     return _dim_check, "Only 2d or 3d kernel supported."
 
+# Build count_used to record params have been converted to attrs
+used_params = {}
```
Comment (kevinthesun) [inline python/tvm/relay/frontend/tensorflow.py]: Only need set?

Comment (kevinthesun) [reply python/tvm/relay/frontend/tensorflow.py]: Agree with @srkreddy1238 if relay build can remove unused params, we don't  need this here.

Comment (lfengad) [reply python/tvm/relay/frontend/tensorflow.py]: Thank you for the suggestions! Then I think the thing to do is simple and we can just remove  the pop operations.

**Comment by lfengad** [comment]
> yes, this is an issue I too see recently while working on TF 2.0 upgrade. Few suggestions on this.
> 
> * I think we can just remove the pop of used param and leave it assuming relay build process removes unused params automatically. cc @tqchen or @jroesch to confirm this behaviour.
> * You may move the test case to default test_forward.py instead of a new file for it.

Thank you for the suggestions! I can try it then!

File: tests/python/frontend/tensorflow/test_sharing_node.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    for device in ["llvm"]:
+        ctx = tvm.context(device, 0)
+        if not ctx.exist:
+            print("Skip because %s is not enabled" % device)
+            continue
+        mod, params = relay.frontend.from_tensorflow(constant_graph,
+                                                     outputs=['output'])
+        with relay.build_config(opt_level=3):
+            graph, lib, params = relay.build(mod,
```
Comment (srkreddy1238) [inline tests/python/frontend/tensorflow/test_sharing_node.py]: You may use the same test case to verify the relay behaviour dropping unused params if frontend just ignores pop. In this case final params after relay build shouldn't contain param '''axis'''

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_sharing_node.py]: Yeah! I think the axis will be removed then. Thank you for the information!

Comment (lfengad) [reply tests/python/frontend/tensorflow/test_sharing_node.py]: Hi, I have already modified the code according to both suggestions from you two! Thank you so much!

File: python/tvm/relay/frontend/tensorflow.py (hunk 1 above)
Comment (lfengad) [reply python/tvm/relay/frontend/tensorflow.py]: Hi, I have already modified the code according to your suggestions! Thank you for the help!

File: tests/python/frontend/tensorflow/test_sharing_node.py (hunk 2 above)
Comment (lfengad) [reply tests/python/frontend/tensorflow/test_sharing_node.py]: Hi I have already done the changes as your suggestions! Could you please update the status of this pr so that it could be merged? Thank you so much!

**Comment by FrozenGene** [comment]
ping @srkreddy1238 Please review it again. If approve, we should follow the instruction to set the status: https://docs.tvm.ai/contribute/code_review.html#approve-and-request-changes-explicitly

**Comment by FrozenGene** [comment]
ping @srkreddy1238

**Comment by FrozenGene** [comment]
Thanks @lfengad's pr and  @srkreddy1238 @kevinthesun 's review. Considering @srkreddy1238 's review comment is fixed and wait @srkreddy1238 7 days / ping two times, additionally two reviewers have approved, so let us merge it firstly. If @srkreddy1238 has other comments, we could update or revert this  pr easily.

**Comment by srkreddy1238** [comment]
Sorry @lfengad. Was a bit disconnected few days over pandemic situations. 
Have no comments and this PR is good. @FrozenGene thanks for merging
